software/bsg_manycore_lib/bsg_mcs_mutex.S

```diff
@@ -0,0 +1,81 @@
+        .text
+        .globl bsg_mcs_mutex_acquire
+bsg_mcs_mutex_acquire:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        sw      x0, 0(a1)                      // lcl->next = 0
+        sw      x0, 4(a1)                      // lcl->unlocked = 0
+        amoswap.w.aq t0, a2, 0(a0)             // predecessor = swap (&mtx, lcl_as_glbl)
+        addi    t1, a1, 4                      // t1 = &lcl->unlocked
+        beqz    t0, bsg_mcs_mutex_acquire_ret  // return if predecessor = 0
+        sw      a2, 0(t0)                      // predecessor->next = lcl_as_glbl
+        lr.w    t0, (t1)                       // unlocked = lcl->unlocked
+        bnez    t0, bsg_mcs_mutex_acquire_ret  // return if unlocked
+bsg_mcs_mutex_acquire_loop:
+        lr.w.aq t0, (t1)                       // unlocked = lcl->unlocked
+        beqz    t0, bsg_mcs_mutex_acquire_loop // while !unlocked
+bsg_mcs_mutex_acquire_ret:
+        ret
+
+        .globl bsg_mcs_mutex_release
+bsg_mcs_mutex_release:
+        // a0 = mtx         : bsg_mcs_mutex_t*
+        // a1 = lcl         : bsg_mcs_mutex_node_t*
+        // a2 = lcl_as_glbl : bsg_mcs_mutex_node_t*
+        li      t1, 1                                       // t1 = 1
+        lw      t0, 0(a1)                                   // next = lcl->next
+        beqz    t0, bsg_mcs_mutex_release_no_successor      // branch if no successor
+        // this is the case where there is a successor
+        // we need only unlock the successor and return
+        fence                                               // fence to implement release semantics
+        sw      t1, 4(t0)                                   // successor->unlocked = 1
+        ret
+bsg_mcs_mutex_release_no_successor:
+        // this is the case where there is no known successor
+        // attempt to swap out the tail pointer with 0
+        amoswap.w.rl t2, x0, 0(a0)                          // victim_tail = swap(&mtx, 0)
+        bne     t2, a2, bsg_mcs_mutex_release_exists_victim // victim_tail == lcl_as_glbl?
+        ret                                                 // there really is no successor -- return
+bsg_mcs_mutex_release_exists_victim:
+        // someone added themselves to the queue and we have removed them
+        // we need to put them back
+        amoswap.w t2, t2, 0(a0)                             // usurper = swap(&mtx, victim_tail)
+bsg_mcs_mutex_release_wait_on_successor:
+        lw      t3, 0(a1)                                   // next = lcl->next
+        bnez    t3, bsg_mcs_mutex_release_wait_on_successor // while (lcl->next != 0)
+        bnez    t2, bsg_mcs_mutex_release_exists_usurper    // was there an usurper?
+        // no usurper exists -- unlock our successor
+        sw      t1, 0(t3)                                   // next->unlocked = 1
+        ret
+bsg_mcs_mutex_release_exists_usurper:       
```

Comment:
Four threads: Releaser, Victim, Usurper, and Naive

What happens if there's a victim, and the naive thread adds itself to the usurper's next before the releaser can fix the usurper's next?